### PR TITLE
Do not consider non-deterministic expressions as invariants in pre-filters

### PR DESCRIPTION
**Pull request by dyemanov**
Related to my prior commit for #1708. Even if the boolean does not access table fields it may still be non-deterministic, in this case it should not be treated as invariant and evaluated just once.

Example:

`select count(*) from employee where rand() > 0.5`

The condition is not correlated with any data but still produces a different result per every invocation. Prior to v5, it returned (approximately) half rows of the result set. In v5, it returns either zero rows or a half of them, depending on what was the result of `RAND()` during its first invocation.

Tested with constant expressions, `RAND()` and SQL functions defined as deterministic and non-deterministic. This patch does not detect determinism too deeply (e.g. whether a field references stream from the current scope or the outer one), only to the level currently needed for the optimizer.

### CODE REVIEW

File: src/jrd/SysFunction.cpp [hunk 1]
```diff
@@ -6962,3 +6962,16 @@ void SysFunction::checkArgsMismatch(int count) const
 		status_exception::raise(Arg::Gds(isc_funmismat) << Arg::Str(name));
 	}
 }
+
+
+bool SysFunction::deterministic() const
+{
+	const MetaName funcName(name);
+
+	return (funcName != "GEN_UUID" &&
```
Comment (asfernandes) [inline src/jrd/SysFunction.cpp:6971]: I think this should be defined in table at `SysFunction::functions`.

**Comment by pavel-zotov** [comment]
Only 5.x is checked now.
FB 6.x has not yet improved, test will fail on it.
Waiting for front-port.

**Comment by dyemanov** [comment]
Frontported to v6.

**Comment by pavel-zotov** [comment]
QA note: removed upper bound for version after this feature was front-ported to 6.x. Checked on 6.0.0.532.

**Pull request merged**